src/c_frontend/tdi_table_key_c.cpp

```diff
@@ -0,0 +1,229 @@
+                                              const uint8_t *value1,
+                                              const uint16_t p_length,
+                                              const size_t size) {
+  auto key = reinterpret_cast<tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueLPM <const uint8_t> keyFieldValue(value1, p_length, size);
+  return key->setValue(field_id, keyFieldValue);
+}
+
+/** Get */
+/* Exact */
+tdi_status_t tdi_key_field_get_value(const tdi_table_key_hdl *key_hdl,
+                                      const tdi_id_t field_id,
+                                      uint64_t *value) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueExact <uint64_t> keyFieldValue(*value);
+  auto sts = key->getValue(field_id, &keyFieldValue);
+  *value = keyFieldValue.value_;
+  return(sts);
+}
+
+tdi_status_t tdi_key_field_get_value_ptr(const tdi_table_key_hdl *key_hdl,
+                                          const tdi_id_t field_id,
+                                          const size_t size,
+                                          uint8_t *value) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueExact <uint8_t> keyFieldValue(value, size);
+  return key->getValue(field_id, &keyFieldValue);
+}
+
+/* string */
+tdi_status_t tdi_key_field_get_value_string_size(
+    const tdi_table_key_hdl *key_hdl,
+    const tdi_id_t field_id,
+    uint32_t *str_size) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  std::string str;
+  tdi::KeyFieldValueExact<std::string> keyFieldValue(str);
+  tdi_status_t status = key->getValue(field_id, &keyFieldValue);
+  *str_size = str.size();
+  return status;
+}
+
+tdi_status_t tdi_key_field_get_value_string(const tdi_table_key_hdl *key_hdl,
+                                            const tdi_id_t field_id,
+                                            char *value) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  std::string str(value);
+  tdi::KeyFieldValueExact<std::string> keyFieldValue(str);
+  tdi_status_t status = key->getValue(field_id, &keyFieldValue);
+  strncpy(value, str.c_str(), str.size());
+  return status;
+}
+
+/* Ternary */
+tdi_status_t tdi_key_field_get_value_and_mask(
+    const tdi_table_key_hdl *key_hdl,
+    const tdi_id_t field_id,
+    uint64_t *value1,
+    uint64_t *value2) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueTernary<uint64_t> keyFieldValue(*value1, *value2);
+  return key->getValue(field_id, &keyFieldValue);
+}
+
+tdi_status_t tdi_key_field_get_value_and_mask_ptr(
+    const tdi_table_key_hdl *key_hdl,
+    const tdi_id_t field_id,
+    const size_t size,
+    uint8_t *value1,
+    uint8_t *value2) {
+  auto key = reinterpret_cast<const tdi::TableKey *>(key_hdl);
+  tdi::KeyFieldValueTernary<unsigned char> keyFieldValue(value1, value2, size);
```

Comment:
I think these will can be uint64 templates. Only the calling ctor will change which is correct here